eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/MessageOnion.scala

```diff
@@ -0,0 +1,131 @@
+   * The reply path uses route blinding, which ensures that the sender doesn't leak its identity to the recipient.
+   */
+  case class ReplyPath(blindedRoute: BlindedRoute) extends OnionMessagePayloadTlv
+
+  /**
+   * Onion messages always use route blinding, even in the forward direction.
+   * This ensures that intermediate nodes can't know whether they're forwarding a message or its reply.
+   * The sender must provide some encrypted data for each intermediate node which lets them locate the next node.
+   */
+  case class EncryptedData(data: ByteVector) extends OnionMessagePayloadTlv
+
+}
+
+object MessageOnion {
+
+  /** Per-hop payload from an onion message (after onion decryption and decoding). */
+  sealed trait PerHopPayload
+
+  /** Per-hop payload for an intermediate node. */
+  case class RelayPayload(records: TlvStream[OnionMessagePayloadTlv]) extends PerHopPayload {
+    val encryptedData: ByteVector = records.get[OnionMessagePayloadTlv.EncryptedData].get.data
+  }
+
+  /** Content of the encrypted data of an intermediate node's per-hop payload. */
+  case class BlindedRelayPayload(records: TlvStream[EncryptedRecipientDataTlv]) {
+    val nextNodeId: PublicKey = records.get[EncryptedRecipientDataTlv.OutgoingNodeId].get.nodeId
+    val nextBlindingOverride: Option[PublicKey] = records.get[EncryptedRecipientDataTlv.NextBlinding].map(_.blinding)
+  }
+
+  /** Per-hop payload for a final node. */
+  case class FinalPayload(records: TlvStream[OnionMessagePayloadTlv]) extends PerHopPayload {
+    val replyPath: Option[OnionMessagePayloadTlv.ReplyPath] = records.get[OnionMessagePayloadTlv.ReplyPath]
+    val encryptedData: ByteVector = records.get[OnionMessagePayloadTlv.EncryptedData].get.data
+  }
+
+  /** Content of the encrypted data of a final node's per-hop payload. */
+  case class BlindedFinalPayload(records: TlvStream[EncryptedRecipientDataTlv]) {
+    val pathId: Option[ByteVector] = records.get[EncryptedRecipientDataTlv.PathId].map(_.data)
+  }
+
+}
+
+object MessageOnionCodecs {
+
+  import MessageOnion._
+  import OnionMessagePayloadTlv._
+  import fr.acinq.eclair.wire.protocol.CommonCodecs._
+  import scodec.codecs._
+  import scodec.{Attempt, Codec}
+
+  private val replyHopCodec: Codec[BlindedNode] = (("nodeId" | publicKey) :: ("encryptedData" | variableSizeBytes(uint16, bytes))).as[BlindedNode]
+
+  private val replyPathCodec: Codec[ReplyPath] = variableSizeBytesLong(varintoverflow, ("firstNodeId" | publicKey) :: ("blinding" | publicKey) :: ("path" | list(replyHopCodec).xmap[Seq[BlindedNode]](_.toSeq, _.toList))).as[BlindedRoute].as[ReplyPath]
+
+  private val encryptedDataCodec: Codec[EncryptedData] = variableSizeBytesLong(varintoverflow, bytes).as[EncryptedData]
+
+  private val onionTlvCodec = discriminated[OnionMessagePayloadTlv].by(varint)
+    .typecase(UInt64(2), replyPathCodec)
+    .typecase(UInt64(10), encryptedDataCodec)
```

Comment:
As seen with Rusty earlier today, we should use 4 here, no need for a gap:
```suggestion
    .typecase(UInt64(4), encryptedDataCodec)
```